src/kde/quick_index.cc

```diff
@@ -0,0 +1,1487 @@
+    // std::partial_sum but I need to think how this would actually look.
+    max_put_buffer_size = 0;
+    for (int rec_proc = 0; rec_proc < rtt_c4::nodes(); rec_proc++) {
+      // calculating the offset SUCKS!!! If anyone can find a better way please help.
+      int offset = 0;
+      for (int send_proc = 0; send_proc < rtt_c4::node(); send_proc++) {
+        if (rec_proc == send_proc)
+          continue;
+        for (size_t bin = 0; bin < nbins; bin++) {
+          if (global_need_bins_per_proc[bin + nbins * rec_proc] > 0) {
+            offset += global_index_per_bin_per_proc[bin + nbins * send_proc];
+          }
+        }
+      }
+      for (auto &map : coarse_index_map) {
+        if (rtt_c4::node() != rec_proc) {
+          size_t gipbpp_index = map.first + nbins * rec_proc;
+          if (global_need_bins_per_proc[gipbpp_index] > 0) {
+            // capture the largest put buffer on this rank
+            if (map.second.size() > max_put_buffer_size)
+              max_put_buffer_size = map.second.size();
+
+            // build up map data
+            put_window_map[map.first].push_back(
+                std::array<int, 3>{rec_proc, proc_ghost_buffer_size[rec_proc], offset});
+            offset += static_cast<int>(map.second.size());
+          }
+        }
+      }
+    }
+
+    // allocate ghost locations
+    local_ghost_locations =
+        std::vector<std::array<double, 3>>(local_ghost_buffer_size, {0.0, 0.0, 0.0});
+    // collect the local ghost locations
+    collect_ghost_data(locations, local_ghost_locations);
+
+  } // End domain decomposed data construction
+}
+
+#ifdef C4_MPI
+//------------------------------------------------------------------------------------------------//
+// call MPI_put using a chunk style write to avoid error in MPI_put with large local buffers.
+auto put_lambda = [](auto &put, auto &put_buffer, auto &put_size, auto &win) {
+  // temporary work around until RMA is available in c4
+  // loop over all ranks we need to send this buffer too.
+  for (auto &putv : put.second) {
+    const int put_rank = putv[0];
+    const int put_rank_buffer_size = putv[1];
+    const int put_offset = putv[2];
+    // This is dumb, but we need to write in chunks because MPI_Put writes
+    // junk with large (>10,000) buffer sizes.
+    int chunk_size = 1000;
+    const auto nchunks = static_cast<int>(
+        std::ceil(static_cast<double>(put_size) / static_cast<double>(chunk_size)));
+    int nput = 0;
+    for (int c = 0; c < nchunks; c++) {
+      chunk_size = std::min(chunk_size, static_cast<int>(put_size) - nput);
+      Check(chunk_size > 0);
+      MPI_Put(&put_buffer[nput], chunk_size, MPI_DOUBLE, put_rank, put_offset, put_rank_buffer_size,
+              MPI_DOUBLE, win);
+      nput += chunk_size;
+    }
+  }
+};
+#endif
+
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief
```

Comment:
Combine this tag with the following paragraph.

Reply:
done.